Metadata API: Implement threshold verification

jku opened the pull request:
#1417 is merged so this is good to go

The delegating Metadata (root or targets) verifies that the delegated metadata is signed by required threshold of keys for the delegated role.

This implementation raises UnsignedMetadataError if threshold is not met. Other possible exceptions are programming errors.

Fixes #1306

Commit:
Metadata API: Implement threshold verification

The delegating Metadata (root or targets) verifies that the delegated
metadata is signed by required threshold of keys for the delegated
role.

Calling the function on non-delegator-metadata or giving a rolename
that is not actually delegated by the delegator is considered a
programming error and ValueError is raised.

If the threshold is not reached, UnsignedMetadataError is raised.

Tweak type annotation of Delegations.keys to match the one for
Root.keys (so they can be assigned to same local variable).

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

sechkova reviewed (approved):
Maybe I've become biased after looking at the code in the next-gen client for a while because I couldn't point out any issues :) 

I was searching for a way to avoid adding `role_name` as an argument of `verify_delegate` but I saw you already opened https://github.com/theupdateframework/tuf/issues/1425 about it so +1 from me on that.

MVrachev commented on tests/test_api.py:
```diff
@@ -352,6 +352,56 @@ def test_metadata_timestamp(self):
+            snapshot.verify_delegate('snapshot', snapshot)
+        # cannot verify with non-existing role
+        with self.assertRaises(ValueError):
+            root.verify_delegate('role1', role1)
```
Nit: If you use a role you haven't used before in the test it will be easier to understand.
Even though it's logical you can't delegate trust from `role1` to `role1`.

jku replied:
This tries to verify that root delegates to role1 (and it does not so ValueError). The comment could be better though
`# try verifying a role that is not delegated by the delegator `

MVrachev replied:
Yes, I didn't read it right that it's the root that's calling it.🤦
Still, I agree that the comment can be improved.

MVrachev reviewed (approved):
I didn't found any issues either.
LGTM!

MVrachev commented on tuf/api/metadata.py:
```diff
@@ -250,59 +250,61 @@ def sign(
+            try:
+                key.verify_signature(delegate, signed_serializer)
+                # keyids are unique. Try to make sure the public keys are too
+                signing_keys.add(key.keyval["public"])
```
Maybe making sure that we have `keyval["public"]` uniqueness can be done instead in `Root` and `Targets`?

jku replied:
yeah possible but I'd say it's a separate issue: #1429

jku replied:
I only very recently realised the TUF spec does not require the existence of "public" at all -- the three keys defined in the spec all happen to have it but it isn't required.

So this line needs to change. The spec does require that KEYIDs are unique so `verify_delegate()` will just respect that and will not try to poke into the internals.

MVrachev commented on tests/test_api.py:
```diff
@@ -359,6 +352,56 @@ def test_metadata_timestamp(self):
+        with self.assertRaises(tuf.exceptions.UnsignedMetadataError):
+            root.verify_delegate('snapshot', snapshot)
+
+        # TODO test higher thresholds
```
Wondering why `higher thresholds` will make a difference?

jku replied:
it's just that we don't have a test here that actually uses a threshold of more than one key successfully. It should work but... it still feels like we're testing the edge case (1 key) when the real problem is N keys.

jku replied:
The fact that I did not add a test case but just the comment does point to a problem in our testing infra: we probably need to be able to produce new signed metadata more easily in the test itself: this is something we should put some effort into (we could use repository_tool... but I don't think we want to even write the metadata to a file, we just want to be able to generate it in memory and test it)

jku replied:
I think the right way forward here is issue #1444

jku replied:
I did end up adding tests for threshold of 2.

jku commented on tuf/api/metadata.py:
```diff
@@ -250,59 +250,61 @@ def sign(
+                roles = self.signed.delegations.roles
+                role = next((r for r in roles if r.name == role_name), None)
+        else:
+            raise ValueError("Call is valid only on delegator metadata")
```
from Trishank: "Why ValueError?"

Valid question as arguments are not the issue here (except if we consider "self" an argument, which it is in a way). 

In any case it's not terribly important as this is a programming error and should not be handled by anyone

jku replied:
I left it a ValueError: I'm not going to create a custom error for something that isn't meant to be caught and can't find a better one in the builtin exceptions.

sechkova replied:
Could be `TypeError` since what `isinstance` checks is the type but is it much better than `ValueError` .. I guess not.

jku replied:
I originally did not pick that because the object type the call is made on (Metadata) is correct ... but I think you're right, it's not perfect but it's better.

jku commented on tuf/api/metadata.py:
```diff
@@ -256,6 +256,61 @@ def sign(
+                keys = self.signed.delegations.keys
+                # Assume role names are unique in delegations.roles: #1426
+                roles = self.signed.delegations.roles
+                role = next((r for r in roles if r.name == role_name), None)
```
Also from trishank: "What does this line do? Looks like magic"
My response was:
> it returns the first role in delegations with name role_name (or None if no match).
> 
> I agree it's not pretty and a bit too clever.
> I could do something like this instead:
> 
> ```
> role = None
> for delegated_role in roles:
>     if delegated_role.name == role_name:
>         role = delegated_role
>         break
> ```
> 
> but that's 5 lines of _initialize-then-modify_ antipattern: not the end of the world but not perfect either.
> 
> the real fix is probably #1426 though (make roles a OrderedDict). Then this code becomes readable and efficient:
> `role = roles.get(role_name)`

I think I'd leave it as is ... but if others think that's too clever I'm happy to change to the more verbose one

joshuagl replied:
Perhaps add a one-liner comment for less familiar readers?
> Find the first role in delegations with name role_name (or None if no match)

MVrachev commented:
In our discussion with @jku  in issue #1438 we agreed that we should make sure that the metadata is accepted even if it has a couple of invalid keys, but it accommodates the requirement placed by threshold.
I think I don't see a test for that. Do you plan to add one @jku?

jku commented:
> In our discussion with @jku in issue #1438 we agreed that we should make sure that the metadata is accepted even if it has a couple of invalid keys, but it accommodates the requirement placed by threshold.
> I think I don't see a test for that. Do you plan to add one @jku?

I think there are two things that we should definitely test:
* test that Key.verify_signature() raises UnsignedMetadataError (or potentially a derived error) if keys are in any way invalid. This  should happen in #1435, I've added the tests that I think make sense there
* test that verify_delegate() handles UnsignedMetadataErrors coming from Key.verify_signature(): catching that error is fine as long as threshold is eventually reached: this should happen in this PR. It does look like there is no test case with failing signatures that should eventually succeed (because threshold is still reached). That would make sense to add (but it does not need to be about invalid keys in particular).

Soo, yes I guess I plan to add one now...

Commit:
tests: Improve verify_delegate() tests

Make sure verify_delegate() succeeds when threshold is reached even if
some signatures fail to verify.

Make sure higher threshold (2/2) works.

Change error type for "Call is valid only on delegator metadata" error.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

jku commented:
rebased on develop, added a commit to add a tests for 
* signature that fails to verify, but overall signatures reach threshold
* threshold that actually contains multiple signatures

This also changes one error type as suggested by teodora

MVrachev reviewed (approved):
Thanks for adding the tests!
Seems good to me!

Commit:
Metadata API: Don't peek into Key internals

There was an attempt at ensuring key content uniqueness in
verify_delegate() by making sure the values corresponding to "public"
keys in Key.keyval dictionaries are unique. This had two issues:
 * it wasn't a security measure: it's not difficult to produce two
   different "public" values of the same key content
 * Spec does not actually guarantee the existence of "public" key in
   the keyval dictionary (the three keys included in the spec just all
   happen to have it)

Luckily the spec does require KEYIDs to be unique so we do not need to
do all this: Just count keyids of keys with verified signatures. Keep
building a Set of keyids as a belt-and-suspenders strategy: Role keyids
are currently guaranteed to be unique but we'd notice here if they
weren't.

Add a logger call for failed verifys: this might useful to figure out
which keys exactly are the issue when a delegate can not be verified.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

jku commented:
Added a commit that removes the attempt to enforce key contents uniqueness:
* I only very recently realised the TUF spec does not require the existence of "public" key within the Key.keyval dictionary at all -- the three keys defined in the spec all happen to have it but it isn't required.
* We know that the attempt is not protection against maliciously crafted keys anyway: creating multiple representations of the same key is not difficult 

Luckily the spec requires that KEYIDs are unique so verify_delegate() will just respect that and will not try to poke into the internals of the keys.

joshuagl commented on tuf/api/metadata.py:
```diff
@@ -273,6 +276,60 @@ def sign(
+        keys for the delegated role 'role_name'.
+
+        Args:
+            role_name: Name of the delegated role to verify
+            delegate: The Metadata object for the delegated role
```
My reading of this is that `verify_delegate` takes an identifier for the delegated role (role_name) and an instance of the delegated role (delegate) – I would expect these two named parameters to have similar names.
* `delegated_role` and `delegated_metadata`?
* `role_name` and `role_metadata`? (handily avoiding the delegate, delegated, delegatee, delegator confusion)

joshuagl reviewed (commented):
This looks good to me. I wonder if the parameter names in `verify_delegate` could be renamed to more clearly indicate that role_name and delegate are related (an identifier and an instance)?

Commit:
Metadata API: verify_delegate: refactor

* Rename arguments so connection between the role name and the
  metadata is stronger.
* Also add a comment on the list comprehension + next() trick.
* Add return value annotation
* Raise early if delegations is None to make the flow more obvious
  (and modify test case so we have coverage for the new case)

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

jku commented:
I chose the more verbose names: it's easy to lose track of where the keys and thresholds really come from and what they are applied on so more verbose seems reasonable.

jku commented:
Someone made mypy rules stricter in the mean time :grimacing: so had to rebase and fix the annotation (no changes in existing commits, only in two new commit).

joshuagl commented on tuf/api/metadata.py:
```diff
@@ -266,6 +269,63 @@ def sign(
+        if role is None:
+            raise ValueError(f"No delegation found for {delegated_role}")
+
+        # verify that delegate is signed by required threshold of unique keys
```
Nit:
```suggestion
        # verify that delegated_metadata is signed by required threshold of unique keys
```

joshuagl commented on tuf/api/metadata.py:
```diff
@@ -266,6 +269,63 @@ def sign(
+            keys = self.signed.keys
+            role = self.signed.roles.get(delegated_role)
+        elif isinstance(self.signed, Targets):
+            if self.signed.delegations:
```
Nit: I wonder if this would be slightly clearer if we did something like:
```Python
if not self.signed.delegations:
    raise TypeError("Targets does not delegate to any other roles")
keys = ...
...
```
?

joshuagl commented on tuf/api/metadata.py:
```diff
@@ -266,6 +269,63 @@ def sign(
+        delegated_metadata: "Metadata",
+        signed_serializer: Optional[SignedSerializer] = None,
+    ) -> None:
+        """Verifies that 'delegate' is signed with the required threshold of
+        keys for the delegated role 'role_name'.
```
```suggestion
        """Verifies that 'delegated_metadata' is signed with the required
        threshold of keys for the 'delegated_role'.
```

joshuagl reviewed (approved):
I really like how easy this is to follow, nice work.

jku commented:
Yes, agree with all suggestions. I've combined all the improvements from suggestions into one commit.

The changes since last push are
* use `if self.signed.delegations is None` to make that path clearer as suggested
* Add a test for this specific case (this required modifying the test data)
* tweak comments as suggested

It's minor enough that I don't think this absolutely needs another approval (so I will merge with the one I have) but ... there's not a lot of code that's more critical for TUF so feel free to have another look

MVrachev reviewed (approved):
I looked at the new changes and it seems good to me!

joshuagl reviewed (approved):
LGTM, thanks!
